src/lib/operators/join_helper/join_output_writing.hpp

```diff
@@ -0,0 +1,278 @@
+ * contains the position list of every segment/chunk in column i.
+ * @param input_table
+ */
+// See usage in _on_execute() for doc.
+inline PosListsByChunk setup_pos_lists_by_chunk(const std::shared_ptr<const Table>& input_table) {
+  Assert(input_table->type() == TableType::References, "Function only works for reference tables");
+
+  struct PosListsHasher {
+    size_t operator()(const PosLists& pos_lists) const { return boost::hash_range(pos_lists.begin(), pos_lists.end()); }
+  };
+
+  std::unordered_map<PosLists, std::shared_ptr<PosLists>, PosListsHasher> shared_pos_lists_by_pos_lists;
+
+  PosListsByChunk pos_lists_by_segment(input_table->column_count());
+  auto pos_lists_by_segment_it = pos_lists_by_segment.begin();
+
+  const auto input_chunks_count = input_table->chunk_count();
+  const auto input_columns_count = input_table->column_count();
+
+  // For every column, for every chunk
+  for (ColumnID column_id{0}; column_id < input_columns_count; ++column_id) {
+    // Get all the input pos lists so that we only have to pointer cast the segments once
+    auto pos_list_ptrs = std::make_shared<PosLists>(input_table->chunk_count());
+    auto pos_lists_iter = pos_list_ptrs->begin();
+
+    // Iterate over every chunk and add the chunks segment with column_id to pos_list_ptrs
+    for (ChunkID chunk_id{0}; chunk_id < input_chunks_count; ++chunk_id) {
+      const auto chunk = input_table->get_chunk(chunk_id);
+      Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
+
+      const auto& ref_segment_uncasted = chunk->get_segment(column_id);
+      const auto ref_segment = std::static_pointer_cast<const ReferenceSegment>(ref_segment_uncasted);
+      *pos_lists_iter = ref_segment->pos_list();
+      ++pos_lists_iter;
+    }
+
+    // pos_list_ptrs contains all position lists of the reference segments for the column_id.
+    auto iter = shared_pos_lists_by_pos_lists.emplace(*pos_list_ptrs, pos_list_ptrs).first;
+
+    *pos_lists_by_segment_it = iter->second;
+    ++pos_lists_by_segment_it;
+  }
+
+  return pos_lists_by_segment;
+}
+
+/**
+ *
+ * @param output_segments [in/out] Vector to which the newly created reference segments will be written.
+ * @param input_table Table which all the position lists reference
+ * @param input_pos_list_ptrs_sptrs_by_segments Contains all position lists to all columns of input table
+ * @param pos_list contains the positions of rows to use from the input table
+ */
+inline void write_output_segments(Segments& output_segments, const std::shared_ptr<const Table>& input_table,
+                                  const PosListsByChunk& input_pos_list_ptrs_sptrs_by_segments,
+                                  std::shared_ptr<RowIDPosList> pos_list) {
+  std::map<std::shared_ptr<PosLists>, std::shared_ptr<RowIDPosList>> output_pos_list_cache;
+
+  // We might use this later, but want to have it outside of the for loop
```

Comment:
Not yours, but this comment is pointless.

Reply:
Thanks for the review.
I guess only reading the changes isn't sufficient. :( I should have seen these ...